File: make_kmz.py
"""
Топо зургийн tile-уудыг Garmin Custom Maps KMZ болгох
Zoom 8 tile-уудыг ашиглана (TMS convention - y inverted)
Max 100 tile/KMZ → олон файлд хуваана
"""
import os, math, zipfile, struct, shutil
from pathlib import Path
from PIL import Image

TILES_DIR = Path("public/tiles/8")
OUT_DIR   = Path("garmin_kmz")
ZOOM      = 8
MAX_TILES = 100  # Garmin Custom Maps limit

OUT_DIR.mkdir(exist_ok=True)

def tile_bounds(x, y_tms, z):
    """TMS tile → (north, south, east, west) lat/lon"""
    n = 2 ** z
    y = n - 1 - y_tms  # TMS → XYZ
    lon_w = x / n * 360 - 180
    lon_e = (x + 1) / n * 360 - 180
    lat_n = math.degrees(math.atan(math.sinh(math.pi * (1 - 2 * y / n))))
    lat_s = math.degrees(math.atan(math.sinh(math.pi * (1 - 2 * (y + 1) / n))))
    return lat_n, lat_s, lon_e, lon_w

def make_kml(tiles_info):
    """tiles_info = list of (img_filename, north, south, east, west)"""
    overlays = ""
    for i, (fname, n, s, e, w) in enumerate(tiles_info):
        overlays += f"""  <GroundOverlay>
    <name>topo_{i}</name>
    <drawOrder>50</drawOrder>
    <Icon><href>files/{fname}</href></Icon>
    <LatLonBox>
      <north>{n:.6f}</north>
      <south>{s:.6f}</south>
      <east>{e:.6f}</east>
      <west>{w:.6f}</west>
    </LatLonBox>
  </GroundOverlay>\n"""
    return f"""<?xml version="1.0" encoding="UTF-8"?>
<kml xmlns="http://www.opengis.net/kml/2.2">
<Document>
  <name>Mongolia Topo 1:200000</name>
{overlays}</Document>
</kml>"""

# Tile бүгдийг цуглуул
all_tiles = []
for x_dir in sorted(TILES_DIR.iterdir()):
    if not x_dir.is_dir(): continue
    x = int(x_dir.name)
    for png in sorted(x_dir.glob("*.png")):
        y_tms = int(png.stem)
        all_tiles.append((x, y_tms, png))

print(f"Нийт {len(all_tiles)} tile олдлоо")

# MAX_TILES-аар хувааж KMZ үүсгэ
chunks = [all_tiles[i:i+MAX_TILES] for i in range(0, len(all_tiles), MAX_TILES)]

for ci, chunk in enumerate(chunks):
    kmz_path = OUT_DIR / f"mongolia_topo_part{ci+1}.kmz"
    tmp_dir  = OUT_DIR / f"_tmp_{ci}"
    tmp_dir.mkdir(exist_ok=True)
    (tmp_dir / "files").mkdir(exist_ok=True)

    tiles_info = []
    for x, y_tms, png_path in chunk:
        n, s, e, w = tile_bounds(x, y_tms, ZOOM)
        jpg_name = f"t_{x}_{y_tms}.jpg"
        # PNG → JPEG (жижигрүүлнэ)
        img = Image.open(png_path).convert("RGB")
        img.save(tmp_dir / "files" / jpg_name, "JPEG", quality=85)
        tiles_info.append((jpg_name, n, s, e, w))

    kml = make_kml(tiles_info)
    (tmp_dir / "doc.kml").write_text(kml, encoding="utf-8")

    with zipfile.ZipFile(kmz_path, "w", zipfile.ZIP_DEFLATED) as zf:
        zf.write(tmp_dir / "doc.kml", "doc.kml")
        for f in (tmp_dir / "files").iterdir():
            zf.write(f, f"files/{f.name}")

    shutil.rmtree(tmp_dir)
    size_mb = kmz_path.stat().st_size / 1024 / 1024
    print(f"  ✅ {kmz_path.name} — {len(chunk)} tile, {size_mb:.1f} MB")

print(f"\nДуусав! {OUT_DIR}/ хавтасноос KMZ файлуудыг")
print("цагны /GARMIN/CustomMaps/ хавтсанд хуулна уу.")

File: make_kmz.js
/**
 * Топо зургийн tile-уудыг Garmin Custom Maps KMZ болгох
 * Zoom 8 tile-уудыг ашиглана (TMS convention)
 * sharp эсвэл jimp-гүй — PNG-г шууд embed хийнэ (Garmin PNG дэмждэг)
 */
const fs   = require('fs');
const path = require('path');
const zlib = require('zlib');

const TILES_DIR = 'public/tiles/8';
const OUT_DIR   = 'garmin_kmz';
const ZOOM      = 8;
const MAX_TILES = 100;

if (!fs.existsSync(OUT_DIR)) fs.mkdirSync(OUT_DIR);

// TMS tile → lat/lon bounds
function tileBounds(x, yTms, z) {
  const n   = Math.pow(2, z);
  const y   = n - 1 - yTms; // TMS → XYZ
  const lonW = x / n * 360 - 180;
  const lonE = (x + 1) / n * 360 - 180;
  const latN = Math.atan(Math.sinh(Math.PI * (1 - 2 * y / n))) * 180 / Math.PI;
  const latS = Math.atan(Math.sinh(Math.PI * (1 - 2 * (y + 1) / n))) * 180 / Math.PI;
  return { n: latN, s: latS, e: lonE, w: lonW };
}

// KML үүсгэх
function makeKml(tilesInfo, partNum) {
  const overlays = tilesInfo.map(({ fname, b }, i) => `  <GroundOverlay>
    <name>topo_p${partNum}_${i}</name>
    <drawOrder>50</drawOrder>
    <Icon><href>files/${fname}</href></Icon>
    <LatLonBox>
      <north>${b.n.toFixed(6)}</north>
      <south>${b.s.toFixed(6)}</south>
      <east>${b.e.toFixed(6)}</east>
      <west>${b.w.toFixed(6)}</west>
    </LatLonBox>
  </GroundOverlay>`).join('\n');

  return `<?xml version="1.0" encoding="UTF-8"?>
<kml xmlns="http://www.opengis.net/kml/2.2">
<Document>
  <name>Mongolia Topo 1:200000 Part ${partNum}</name>
${overlays}
</Document>
</kml>`;
}

// ZIP/KMZ үүсгэх (node built-in зипгүй тул archiver ашиглах)
// Эхлээд archiver байгаа эсэх шалгах
let useArchiver = false;
try { require.resolve('archiver'); useArchiver = true; } catch {}

if (!useArchiver) {
  console.log('📦 archiver суулгаж байна...');
  require('child_process').execSync('npm install archiver --save-dev', { stdio: 'inherit' });
  useArchiver = true;
}

const { ZipArchive } = require('archiver');

// Tile бүгдийг цуглуул
const allTiles = [];
const xDirs = fs.readdirSync(TILES_DIR).sort((a, b) => +a - +b);
for (const xName of xDirs) {
  const xPath = path.join(TILES_DIR, xName);
  if (!fs.statSync(xPath).isDirectory()) continue;
  const x = parseInt(xName);
  const pngs = fs.readdirSync(xPath).filter(f => f.endsWith('.png')).sort((a, b) => +a.replace('.png','') - +b.replace('.png',''));
  for (const png of pngs) {
    const yTms = parseInt(png);
    allTiles.push({ x, yTms, pngPath: path.join(xPath, png) });
  }
}

console.log(`Нийт ${allTiles.length} tile олдлоо`);

// MAX_TILES-аар хуваах
const chunks = [];
for (let i = 0; i < allTiles.length; i += MAX_TILES) {
  chunks.push(allTiles.slice(i, i + MAX_TILES));
}

async function makeKmz(chunk, partNum) {
  return new Promise((resolve, reject) => {
    const kmzPath = path.join(OUT_DIR, `mongolia_topo_part${partNum}.kmz`);
    const output  = fs.createWriteStream(kmzPath);
    const archive = new ZipArchive({ zlib: { level: 6 } });

    output.on('close', () => {
      const mb = (archive.pointer() / 1024 / 1024).toFixed(1);
      console.log(`  ✅ mongolia_topo_part${partNum}.kmz — ${chunk.length} tile, ${mb} MB`);
      resolve();
    });
    archive.on('error', reject);
    archive.pipe(output);

    // Tile файлуудыг нэмэх
    const tilesInfo = [];
    for (const { x, yTms, pngPath } of chunk) {
      const b     = tileBounds(x, yTms, ZOOM);
      const fname = `t_${x}_${yTms}.png`;
      archive.file(pngPath, { name: `files/${fname}` });
      tilesInfo.push({ fname, b });
    }

    // KML нэмэх
    const kml = makeKml(tilesInfo, partNum);
    archive.append(kml, { name: 'doc.kml' });
    archive.finalize();
  });
}

(async () => {
  for (let i = 0; i < chunks.length; i++) {
    await makeKmz(chunks[i], i + 1);
  }
  console.log(`\n✅ Дуусав! ${chunks.length} KMZ файл → ${OUT_DIR}/ хавтаст байна`);
  console.log('📲 Цагны /GARMIN/CustomMaps/ хавтсанд хуулна уу.');
})();
